### DISCUSSION TITLE
Is there a way to know which items are selected in a DataGrid?

### DISCUSSION THREAD

**Discussion by IronEaterBeast** [AvaloniaUI/Avalonia#15306] [Q&A]
The DataGrid is already bound to a variable called AllApiInfo. When I click on certain rows in the window, the color changes to indicate selection, but I don't know how to get these selected items in C#.

`DataGrid ItemsSource="{Binding AllApiInfo}"`

![2024-04-10 154431](https://github.com/AvaloniaUI/Avalonia/assets/9688223/f05f59de-3cb2-4c6b-aa0b-dc68d3ae965d)

**Comment by timunie** [comment]
similar to WPF, you can't for DataGrid bind to it. You'll need a Behavior or code behind here.

**Reply by IronEaterBeast** [reply]
After I bind the Datagrid to AllApiInfo in XAML, how can I get selected items  in code-behind? I haven't worked with WPF before, my first c# program is Avalonia.

**Reply by timunie** [reply]
```
<DataGrid x:Name="MyGrid" />

// in code-behind:

MyGrid.SelectedItems ...
```

**Reply by IronEaterBeast** [reply]
I have another question. 

I couldn't find how to implement the deselect (or toggle selection) functionality in a DataGrid. I can use datagridname.SelectAll() to select all items and datagridname.SelectedItems.Clear() to clear all selections, but I don't know how to implement deselecting individual rows. It seems like I don't have a method to iterate through all rows, check if a row is selected, and then toggle its selection status.

Any suggestions?

**Reply by timunie** [reply]
Maybe TreeDataGrid is more modern and the better control for you here? DataGrid is one of the most complex controls to maintain so some things other controls can do, DataGrid can't by design. #8060

**Reply by thevortexcloud** [reply]
> , you can't for DataGrid bind to it

From memory: I have bound SelectedItems before in a datagrid without issue before. The only real problem was there is not a straightforwards way to get notified about selection changes. But you can work around that with various custom collections. Granted that was a long time ago so I am possibly misremembering.
